File: game_of_sim/sim.cpp
#include <bits/stdc++.h>

using namespace std;

class Sim {
private:
    int onMove = 1;
    bool finished = false;
    vector<string> edge_labels;
    vector<string> R, RvB, B, moves;
    map<string, int> edge_colors; // 0 not colored yet, 1 red, 2 blue
    string move10 = "", move12 = "";
    void normalizeEdge(string &edge) {
        if (edge.length() != 2 || edge[0]==edge[1]
            || edge[0]<'A' || edge[0]>'F'
            || edge[1]<'A' || edge[1]>'F')
            throw invalid_argument("Invalid edge label");
        if (edge[0] > edge[1]) swap(edge[0], edge[1]);
    }
    string makeEdge(char P, char Q) {
        auto ans = string(1, P) + Q;
        normalizeEdge(ans);
        return ans;
    }
    bool edgeFromVertex(string PQ, char P) {
        return PQ[0] == P || PQ[1] == P;
    }
    char commonVertex(string PQ, string PR) {
        if (PQ[0] == PR[0]) return PQ[0];
        if (PQ[0] == PR[1]) return PQ[0];
        if (PQ[1] == PR[0]) return PQ[1];
        if (PQ[1] == PR[1]) return PQ[1];
        throw invalid_argument("Invalid labels");
    }
    bool checkGameFinished() {
        for(auto &PQ: edge_labels) {
            if (edge_colors[PQ] == 0) continue;
            int colorPQ = getEdgeColor(PQ);
            for(char R = 'A'; R<='F'; R++) {
                if (R==PQ[0] || R==PQ[1]) continue;
                string PR = makeEdge(PQ[0], R);
                string QR = makeEdge(PQ[1], R);
                if (getEdgeColor(PR)==colorPQ && getEdgeColor(QR)==colorPQ) {
                    finished = true;
                    return finished;
                }
            }
        }
        return finished;
    }

    bool isPlayable(string PQ, int color) {
        if (edge_colors[PQ]) return false;
        for(char R = 'A'; R<='F'; R++) {
            if (R==PQ[0] || R==PQ[1]) continue;
            string PR = makeEdge(PQ[0], R);
            string QR = makeEdge(PQ[1], R);
            if (getEdgeColor(PR)==color && getEdgeColor(QR)==color) return false;
        }
        return true;
    }

    void computePlayableEdges() {
        R.clear(); B.clear(); RvB.clear();
        for(auto &PQ: edge_labels) {
            if (edge_colors[PQ] > 0) continue;
            bool redPlayable = isPlayable(PQ, 1);
            bool bluePlayable = isPlayable(PQ, 2);
            if (redPlayable) {
                if (bluePlayable) RvB.push_back(PQ);
                else R.push_back(PQ);
            } else if (bluePlayable) B.push_back(PQ);
        }
    }

    vector<string> computeMinX() {
        map<string, int> Xs;
        for(auto QS: RvB) {
            edge_colors[QS] = 2;
            Xs[QS] = 0;
            for(auto PQ: RvB) {
                if (PQ != QS && !isPlayable(PQ, 2)) Xs[QS]++;//PQ is lost up on playing on QS
            }
            for(auto PQ: B) {
                if (PQ != QS && !isPlayable(PQ, 2)) Xs[QS]++;//PQ is lost up on playing on QS
            }
            edge_colors[QS] = 0;
        }
        vector<pair<int, string>> X;
        for(auto [edge, val]: Xs) X.emplace_back(val, edge);
        sort(X.begin(), X.end());
        // debug(X);
        vector<string> ans; ans.push_back(X[0].second);
        for(int i=1; i<X.size(); i++) {
            if (X[i].first == X[0].first) ans.push_back(X[i].second);
        }
        return ans;
    }

    int computeZ(string ij) {
        int ans = 0;
        for(char k='A'; k<='F'; k++) {
            if (k==ij[0] || k==ij[1]) continue;
            string ik = makeEdge(ij[0], k);
            string jk = makeEdge(ij[1], k);
            if (isPlayable(ik, 2) && isPlayable(jk, 2)) ans+=2;
            else if (edge_colors[ik] == 1 && isPlayable(jk, 1)) ans++;
            else if (edge_colors[jk] == 1 && isPlayable(ik, 1)) ans++;
        }
        return ans;
    }

    string getBestMove() {
        computePlayableEdges();
        vector<string> Xs = computeMinX();
        if (Xs.size() == 1) return Xs[0];
        vector<pair<int, string>> Zs;
        for(string ij: Xs) {
            int z=computeZ(ij);
            Zs.emplace_back(z, ij);
        }
        sort(Zs.begin(), Zs.end());
        // debug(Zs);
        return Zs[0].second;
    }

public:
    Sim() {
        reset();
    }

    void reset() {
        onMove = 1;
        moves.clear();
        finished = false;
        if (edge_labels.empty()) {
            for(char P='A'; P<'F'; P++) {
                for(char Q=P+1; Q<='F'; Q++) {
                    edge_labels.push_back(makeEdge(P, Q));
                }
            }
        }
        edge_colors.clear();
        for(auto &label: edge_labels) edge_colors[label] = 0;
    }

    void play(string PQ) {
        normalizeEdge(PQ);
        if (finished) throw logic_error("The game has been finished");
        if (getEdgeColor(PQ) > 0) throw logic_error("This move is already played");
        moves.push_back(PQ);
        edge_colors[PQ] = 2 - (onMove%2);
        onMove++;
        checkGameFinished();
        if (useEngine && !finished && isBlueTurn()) {//machine to play
            if (onMove == 2) { // Rule 1
                for(char R='A'; R<'F'; R++) {
                    if (R != PQ[0] && R != PQ[1]) {
                        play(makeEdge(PQ[0], R));
                        return;
                    }
                }
            } else if (onMove == 4) { // Rule 2
                set<char> st;
                for(auto &edge: moves) {
                    st.insert(edge[0]); st.insert(edge[1]);
                }
                if (st.size() == 3) { //3 first moves make a triangle
                    string PR = moves[0];
                    char P = commonVertex(PR, PQ);
                    for(int i=edge_labels.size()-1; i>=0; i--) {
                        if (!edge_colors[edge_labels[i]] && edgeFromVertex(edge_labels[i], P)) {
                            play(edge_labels[i]);
                            return;
                        }
                    }
                } else {
                    char V = commonVertex(moves[0], moves[1]);
                    char R = moves[0][0] == V ? moves[0][1] : moves[0][0];
                    char S = moves[1][0] == V ? moves[1][1] : moves[1][0];
                    play(makeEdge(R, S));
                    return;
                }
            } else if (onMove == 6 || onMove == 8) {
                string bestMove = getBestMove(); // Rule 3
                play(bestMove);
                return;
            } else if (onMove == 10) {
                move10 = move12 = "";
                computePlayableEdges();
                if (R.size() == 1 && RvB.size() == 2 && B.size() == 3) { //Rule 4
                    for(auto side1: RvB) {
                        edge_colors[side1] = 2;
                        for(auto side2: B) {
                            if (isPlayable(side2, 2)) {
                                edge_colors[side2] = 2;
                                for(auto side3: B) {
                                    if (side3 != side2 && isPlayable(side3, 2)) {
                                        move10 = side1; move12 = side2;
                                        break;
                                    }
                                }
                                edge_colors[side2] = 0;
                            }
                        }
                        edge_colors[side1] = 0;
                    }
                    play(move10);
                    return;
                } else if (R.size() == 1 && RvB.size() == 3 && B.size() == 2) { //Rule 5: "AF", "AB", "BF", "EF", "CD", "CF", "BE", "AE", "DF"
                    for(auto side1: RvB) {
                        edge_colors[side1] = 2;
                        bool rule5_best_move = true;
                        for(auto side2: RvB) {
                            if (side2 == side1) continue;
                            if (!isPlayable(side2, 2)) {
                                rule5_best_move = false;
                                break;
                            }
                        }
                        edge_colors[side1] = 0;
                        if (rule5_best_move) {
                            play(side1);
                            return;
                        }
                    }
                    throw logic_error("Strategy not found on move " + to_string(onMove));
                } else {
                    // PrintAnalysis();
                    string bestMove = getBestMove(); // Rule 3
                    play(bestMove);
                    return;
                }
            } else if (onMove == 12) {
                if (move12 != "") { // Rule 4
                    play(move12);
                } else {
                    computePlayableEdges();
                    // PrintAnalysis();
                    if (!RvB.empty()) {
                        play(RvB[0]);
                    } else if (!B.empty()) {
                        play(B[0]);
                    } else {
                        // PrintAnalysis();
                        throw logic_error("Strategy not found on move " + to_string(onMove));
                    }
                }
                return;
            } else if (onMove == 14) {
                computePlayableEdges();
                if (!RvB.empty()) {
                    play(RvB[0]);
                } else if (!B.empty()) {
                    play(B[0]);
                } else {
                    // PrintAnalysis();
                    throw logic_error("Strategy not found on move " + to_string(onMove));
                }
            }
        }
    }

    bool useEngine = true;
    vector<string> getMoves() {return moves;}
    bool isRedTurn() { return onMove % 2 == 1; }
    bool isBlueTurn() { return onMove % 2 == 0; }
    bool isFinished() { return finished; }
    int getEdgeColor(string edge) {
        normalizeEdge(edge);
        return edge_colors[edge];
    }
    int getEdgeColor(char P, char Q) {
        string edge = makeEdge(P, Q);
        return getEdgeColor(edge);
    }
    vector<string> getTriangleEdges() {
        if (!finished) return {};
        for(char P='A'; P<='F'; P++) {
            for(char Q=P+1; Q<='F'; Q++) {
                if (!getEdgeColor(P, Q)) continue;
                for(char S=Q+1; S<='F'; S++) {
                    if (getEdgeColor(P, Q) == getEdgeColor(Q, S) && getEdgeColor(P, Q) == getEdgeColor(P, S)) {
                        return {makeEdge(P, Q), makeEdge(Q, S), makeEdge(S, P)};
                    }
                }
            }
        }
        return {};
    }

    // void PrintAnalysis() {
    //     debug(getMoves());
    //     computePlayableEdges();
    //     debug(R);
    //     debug(RvB);
    //     debug(B);
    // }
};

signed main() {
    Sim sim;
    try {
        sim.play("AF");
        sim.play("FB");
        sim.play("CD");
        sim.play("EB");
        sim.play("DF");
        sim.play("AC");
        sim.play("CE");
        // sim.PrintAnalysis();
        // if (sim.isFinished()) {
        //     debug(sim.getTriangleEdges());
        // }
    }
    catch(const invalid_argument &e) {
        cerr << "Error: " << e.what() << "\n";
    }
    catch(const logic_error &e) {
        cerr << "Error: " << e.what() << "\n";
    }
    return 0;
}

File: game_of_sim/sim.py
from typing import List, Dict, Set

class Sim:
    def __init__(self):
        self.__onMove = 1
        self.__finished = False
        self.__edge_labels = []
        self.__R, self.__RvB, self.__B, self.__moves = [], [], [], []
        self.__edge_colors = {}  # 0 not colored yet, 1 red, 2 blue
        self.__move10, self.__move12 = "", ""
        self.use_engine = True
        self.reset()

    def __normalize_edge(self, edge):
        if len(edge) != 2 or edge[0] == edge[1] or edge[0] < 'A' or edge[0] > 'F' or edge[1] < 'A' or edge[1] > 'F':
            raise ValueError("Invalid edge label")
        if edge[0] > edge[1]:
            edge = edge[1] + edge[0]
        return edge

    def __make_edge(self, P, Q):
        ans = P + Q
        return self.__normalize_edge(ans)

    def edge_from_vertex(self, PQ, P):
        return PQ[0] == P or PQ[1] == P

    def __common_vertex(self, PQ, PR):
        if PQ[0] == PR[0]:
            return PQ[0]
        if PQ[0] == PR[1]:
            return PQ[0]
        if PQ[1] == PR[0]:
            return PQ[1]
        if PQ[1] == PR[1]:
            return PQ[1]
        raise ValueError("Invalid labels")

    def __check_game_finished(self):
        for PQ in self.__edge_labels:
            if self.__edge_colors[PQ] == 0:
                continue
            color_PQ = self.get_edge_color(PQ)
            for R in 'ABCDEF':
                if R == PQ[0] or R == PQ[1]:
                    continue
                PR = self.__make_edge(PQ[0], R)
                QR = self.__make_edge(PQ[1], R)
                if self.get_edge_color(PR) == color_PQ and self.get_edge_color(QR) == color_PQ:
                    self.__finished = True
                    return self.__finished
        return self.__finished

    def __is_playable(self, PQ, color):
        if self.__edge_colors[PQ]:
            return False
        for R in 'ABCDEF':
            if R == PQ[0] or R == PQ[1]:
                continue
            PR = self.__make_edge(PQ[0], R)
            QR = self.__make_edge(PQ[1], R)
            if self.get_edge_color(PR) == color and self.get_edge_color(QR) == color:
                return False
        return True

    def __compute_playable_edges(self):
        self.__R.clear()
        self.__B.clear()
        self.__RvB.clear()
        for PQ in self.__edge_labels:
            if self.__edge_colors[PQ] > 0:
                continue
            red_playable = self.__is_playable(PQ, 1)
            blue_playable = self.__is_playable(PQ, 2)
            if red_playable:
                if blue_playable:
                    self.__RvB.append(PQ)
                else:
                    self.__R.append(PQ)
            elif blue_playable:
                self.__B.append(PQ)

    def __compute_min_x(self):
        Xs = {}
        for QS in self.__RvB:
            self.__edge_colors[QS] = 2
            Xs[QS] = 0
            for PQ in self.__RvB:
                if PQ != QS and not self.__is_playable(PQ, 2):
                    Xs[QS] += 1  # PQ is lost upon playing on QS
            for PQ in self.__B:
                if PQ != QS and not self.__is_playable(PQ, 2):
                    Xs[QS] += 1  # PQ is lost upon playing on QS
            self.__edge_colors[QS] = 0
        X = sorted(Xs.items(), key=lambda x: x[1])
        min_x_val = X[0][1]
        ans = [edge for edge, val in X if val == min_x_val]
        return ans

    def __compute_z(self, ij):
        ans = 0
        for k in 'ABCDEF':
            if k == ij[0] or k == ij[1]:
                continue
            ik = self.__make_edge(ij[0], k)
            jk = self.__make_edge(ij[1], k)
            if self.__is_playable(ik, 2) and self.__is_playable(jk, 2):
                ans += 2
            elif self.__edge_colors[ik] == 1 and self.__is_playable(jk, 1):
                ans += 1
            elif self.__edge_colors[jk] == 1 and self.__is_playable(ik, 1):
                ans += 1
        return ans

    def __get_best_move(self):
        self.__compute_playable_edges()
        Xs = self.__compute_min_x()
        if len(Xs) == 1:
            return Xs[0]
        Zs = [(self.__compute_z(ij), ij) for ij in Xs]
        Zs.sort()
        return Zs[0][1]

    def reset(self):
        self.__onMove = 1
        self.__moves.clear()
        self.__finished = False
        if not self.__edge_labels:
            for P in 'ABCDEF':
                for Q in 'ABCDEF':
                    if P < Q:
                        self.__edge_labels.append(self.__make_edge(P, Q))
        self.__edge_colors.clear()
        for label in self.__edge_labels:
            self.__edge_colors[label] = 0

    def play(self, PQ):
        PQ = self.__normalize_edge(PQ)
        if self.__finished:
            return
            raise ValueError("The game has been finished")
        if self.get_edge_color(PQ) > 0:
            return
            raise ValueError("This move is already played")
        self.__moves.append(PQ)
        self.__edge_colors[PQ] = 2 - (self.__onMove % 2)
        self.__onMove += 1
        self.__check_game_finished()

        if not self.__finished and self.use_engine and self.is_blue_turn():
            if self.__onMove == 2:
                for R in 'ABCDEF':
                    if R != PQ[0] and R != PQ[1]:
                        self.play(self.__make_edge(PQ[0], R))
                        return
            elif self.__onMove == 4:
                st = set()
                for edge in self.__moves:
                    st.add(edge[0])
                    st.add(edge[1])
                if len(st) == 3:
                    PR = self.__moves[0]
                    P = self.__common_vertex(PR, PQ)
                    for i in range(len(self.__edge_labels) - 1, -1, -1):
                        if not self.__edge_colors[self.__edge_labels[i]] and self.edge_from_vertex(self.__edge_labels[i], P):
                            self.play(self.__edge_labels[i])
                            return
                else:
                    V = self.__common_vertex(self.__moves[0], self.__moves[1])
                    R = self.__moves[0][0] if self.__moves[0][0] != V else self.__moves[0][1]
                    S = self.__moves[1][0] if self.__moves[1][0] != V else self.__moves[1][1]
                    self.play(self.__make_edge(R, S))
                    return
            elif self.__onMove == 6 or self.__onMove == 8:
                best_move = self.__get_best_move()
                self.play(best_move)
                return
            elif self.__onMove == 10:
                self.__move10, self.__move12 = "", ""
                self.__compute_playable_edges()
                if len(self.__R) == 1 and len(self.__RvB) == 2 and len(self.__B) == 3:
                    for side1 in self.__RvB:
                        self.__edge_colors[side1] = 2
                        for side2 in self.__B:
                            if self.__is_playable(side2, 2):
                                self.__edge_colors[side2] = 2
                                for side3 in self.__B:
                                    if side3 != side2 and self.__is_playable(side3, 2):
                                        self.__move10 = side1
                                        self.__move12 = side2
                                        break
                                self.__edge_colors[side2] = 0
                        self.__edge_colors[side1] = 0
                    self.play(self.__move10)
                    return
                elif len(self.__R) == 1 and len(self.__RvB) == 3 and len(self.__B) == 2: # Rule 5: "AF", "AB", "BF", "EF", "CD", "CF", "BE", "AE", "DF"
                    for side1 in self.__RvB:
                        self.__edge_colors[side1] = 2
                        rule5_best_move = True
                        for side2 in self.__RvB:
                            if side1 == side2:
                                continue
                            if not self.__is_playable(side2, 2):
                                rule5_best_move = False
                                break
                        self.__edge_colors[side1] = 0
                        if rule5_best_move:
                            self.play(side1)
                            return
                    raise ValueError(f"Strategy not found on move {self.__onMove}")
                else:
                    best_move = self.__get_best_move()
                    self.play(best_move)
                    return
            elif self.__onMove == 12:
                if self.__move12 != "":
                    self.play(self.__move12)
                else:
                    self.__compute_playable_edges()
                    if self.__RvB:
                        self.play(self.__RvB[0])
                    elif self.__B:
                        self.play(self.__B[0])
                    else:
                        raise ValueError(f"Strategy not found on move {self.__onMove}")
                return
            elif self.__onMove == 14:
                self.__compute_playable_edges()
                if self.__RvB:
                    self.play(self.__RvB[0])
                elif self.__B:
                    self.play(self.__B[0])
                else:
                    raise ValueError(f"Strategy not found on move {self.__onMove}")

    def get_moves(self):
        return self.__moves

    def get_next_move(self):
        return self.__onMove

    def is_red_turn(self):
        return self.__onMove % 2 == 1

    def is_blue_turn(self):
        return self.__onMove % 2 == 0

    def is_finished(self):
        return self.__finished

    def get_all_edge_labels(self):
        return self.__edge_labels

    def get_edge_color(self, edge):
        edge = self.__normalize_edge(edge)
        return self.__edge_colors[edge]

    def get_line_color(self, P, Q):
        edge = self.__make_edge(P, Q)
        return self.get_edge_color(edge)

    def get_triangle_edges(self):
        if not self.__finished:
            return []
        for P in 'ABCDEF':
            for Q in 'ABCDEF':
                if P < Q and self.get_line_color(P, Q):
                    for S in 'ABCDEF':
                        if S > Q and self.get_line_color(P, Q) == self.get_line_color(Q, S) == self.get_line_color(P, S):
                            return [self.__make_edge(P, Q), self.__make_edge(Q, S), self.__make_edge(S, P)]
        return []
